cedar-policy-validator/src/schema.rs

```diff
@@ -971,6 +949,134 @@ fn internal_name_to_entity_type(
     Name::try_from(name).map(Into::into)
 }
 
+/// Holds the sets of all entity type, common type, and action definitions
+/// (fully-qualified names) in all fragments.
+#[derive(Debug)]
+pub struct AllDefs {
+    /// All entity type definitions, in all fragments, as fully-qualified names.
+    entity_defs: HashSet<InternalName>,
+    /// All common type definitions, in all fragments, as fully-qualified names.
+    common_defs: HashSet<InternalName>,
+    /// All action definitions, in all fragments, with fully-qualified typenames.
+    action_defs: HashSet<EntityUID>,
+}
+
+impl AllDefs {
+    /// Build the sets of all entity type, common type, and action definitions
+    /// (fully-qualified names) in all fragments.
+    pub fn new<'a, N: 'a, A: 'a, I>(fragments: impl Fn() -> I) -> Self
+    where
+        I: Iterator<Item = &'a ValidatorSchemaFragment<N, A>>,
+    {
+        Self {
+            entity_defs: fragments()
+                .flat_map(|f| f.0.iter())
+                .flat_map(|ns_def| ns_def.all_declared_entity_type_names().cloned())
+                .collect(),
+            common_defs: fragments()
+                .flat_map(|f| f.0.iter())
+                .flat_map(|ns_def| ns_def.all_declared_common_type_names().cloned())
+                .collect(),
+            action_defs: fragments()
+                .flat_map(|f| f.0.iter())
+                .flat_map(|ns_def| ns_def.all_declared_action_names().cloned())
+                .collect(),
+        }
+    }
+
+    /// Build an [`AllDefs`] assuming that the given fragment is the only
+    /// fragment that exists.
+    /// Any names referring to definitions in other fragments will not resolve
+    /// properly.
+    #[cfg(test)]
+    pub fn single_fragment<N, A>(fragment: &ValidatorSchemaFragment<N, A>) -> Self {
+        Self::new(|| std::iter::once(fragment))
+    }
+
+    /// Is the given (fully-qualified) [`InternalName`] defined as an entity
+    /// type in any fragment?
+    pub fn is_defined_as_entity(&self, name: &InternalName) -> bool {
+        self.entity_defs.contains(name)
+    }
+
+    /// Is the given (fully-qualified) [`InternalName`] defined as a common type
+    /// in any fragment?
+    pub fn is_defined_as_common(&self, name: &InternalName) -> bool {
+        self.common_defs.contains(name)
+    }
+
+    /// Is the given (fully-qualified) [`EntityUID`] defined as an action in any
+    /// fragment?
+    pub fn is_defined_as_action(&self, euid: &EntityUID) -> bool {
+        self.action_defs.contains(euid)
+    }
+
+    /// Mark the given [`InternalName`] as defined as a common type
+    pub fn mark_as_defined_as_common_type(&mut self, name: InternalName) {
+        self.common_defs.insert(name);
+    }
+
+    /// Return an error if the definitions in this [`AllDefs`] violate the
+    /// restrictions specified in
+    /// [RFC 70](https://github.com/cedar-policy/rfcs/blob/main/text/0070-disallow-empty-namespace-shadowing.md)
```

Comment:
Could this be named something more self explanatory? I suppose the details of the checks are a bit subtle too express in the function name, but I'd at least like the RFC summary to be included in the comment.

Reply:
Renamed to `rfc_70_shadowing_checks` and expanded the doc comment here.